IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs

```diff
@@ -0,0 +1,281 @@
+                Task.Delay(500).Wait();
+                try
+                {
+                    if (func())
+                    {
+                        return true;
+                    }
+                }
+                catch (WebDriverTimeoutException)
+                {
+                    return false;
+                }
+
+            } while (DateTime.Now <= end);
+            return false;
+        }
+
+        public bool IsElementDisplayedAndExisting(By by, int secondsToWait = 15)
+        {
+            // Now that we have a WebElement.Initialized property, see if it makes sense to refactor this logic to look at that instead?
+            // Or is it worth still re-trying, just in case the site was in a bad state prior to calling this?
+            // If the latter, the default secondsToWait can be kicked down to maybe 5.
+            WebDriverWait wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(secondsToWait));
+            try
+            {
+                wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.ElementExists(by));
+            }
+            catch (WebDriverTimeoutException) { }
+            IReadOnlyCollection<IWebElement> elements = Driver.FindElements(by);
+            if (elements.Count > 0)
+            {
+                return elements.ElementAt(0).Displayed;
+            }
+            return false;
+        }
+
+        public void FrameSwitchAttempt(params By[] bys)
+        {
+            Exception ex = new Exception();
+            for (int i = 0; i < 50; i++)
+            {
+                try
+                {
+                    Driver.SwitchTo().DefaultContent();
+                    ex = null;
+                    foreach (By by in bys)
+                    {
+                        // Don't use our WebElement extension for this as it has trouble being casted to IWebElementReference
+                        IWebElement element = Driver.FindElement(by);
+                        Console.WriteLine("Switching to frame " + by);
+                        Driver.SwitchTo().Frame(element);
+                    }
+                    break;
+                }
+                // Try again.
+                catch (NoSuchFrameException e)
+                {
+                    ex = e;
+                }
+                catch (InvalidOperationException e)
+                {
+                    ex = e;
+                }
+                catch (StaleElementReferenceException e)
+                {
+                    ex = e;
+                }
+                catch (NotFoundException e)
+                {
+                    ex = e;
+                }
+                Task.Delay(500).Wait();
+            }
+            if (ex != null)
+                throw ex;
+            Task.Delay(500).Wait();
+        }
+
+        public Boolean SwitchWindow(string title)
+        {
+            Exception ex = new Exception();
```

Comment:
Exception variable appears to be unused.

Reply:
This is one of the bigger "rough" things in here. I am going to refactor this method, FrameSwitchATtempt() and Alert() to conform more to the retrying for a specified period of time pattern instead of a number of attempts. I'll remove the reference (or use it) when I refactor the method.